symengine/functions.cpp

```diff
@@ -2895,6 +2895,159 @@ RCP<const Basic> uppergamma(const RCP<const Basic> &s, const RCP<const Basic> &x
+{
+    std::size_t seed = BETA;
+    hash_combine<Basic>(seed, *x_);
+    hash_combine<Basic>(seed, *y_);
+    return seed;
+}
+
+bool Beta::__eq__(const Basic &o) const
+{
+    if (is_a<Beta>(o) and
+        eq(*x_, *(static_cast<const Beta &>(o).x_)) and
+        eq(*y_, *(static_cast<const Beta &>(o).y_)))
+        return true;
+    return false;
+}
+
+int Beta::compare(const Basic &o) const
+{
+    SYMENGINE_ASSERT(is_a<Beta>(o))
+    const Beta &lg = static_cast<const Beta &>(o);
+    if (eq(*x_, *(lg.x_))) {
+        return y_->__cmp__(*(static_cast<const Beta &>(o).y_));
+    }
+    if (eq(*x_, *(lg.y_))) {
+        return y_->__cmp__(*(static_cast<const Beta &>(o).x_));
+    }
+    if (eq(*y_,*(lg.y_))) {
+        return x_->__cmp__(*(static_cast<const Beta&>(o).x_));
+    }
+    if (eq(*y_,*(lg.x_))) {
+        return x_->__cmp__(*(static_cast<const Beta&>(o).y_));
+    }
+    throw std::runtime_error("Not yet implemented");
+}
+
+RCP<const Basic> Beta::rewrite_as_gamma() const
+{
+    return div(mul(gamma(x_),gamma(y_)),sub(add(x_,y_),one));
+}
+inline RCP<const Basic> gamma_positive_int(const RCP<const Basic> &arg)
+{
+    SYMENGINE_ASSERT(is_a<Integer>(*arg))
+    RCP<const Integer> arg_ = rcp_static_cast<const Integer>(arg);
+    SYMENGINE_ASSERT(arg_->is_positive())
+    return factorial((arg_->subint(*one))->as_int());
+}
+inline RCP<const Basic> gamma_multiple_2(const RCP<const Basic>& arg)
+{
+    SYMENGINE_ASSERT(is_a<Rational>(*arg))
+    RCP<const Rational> arg_ = rcp_static_cast<const Rational>(arg);
+    SYMENGINE_ASSERT(arg_->i.get_den() == 2)
+    RCP<const Integer> n, k;
+    RCP<const Number> coeff;
+    n = quotient_f(*(integer(abs(arg_->i.get_num()))), *(integer(arg_->i.get_den())));
+    if (arg_->is_positive()) {
+        k = n;
+        coeff = one;
+    } else {
+        n = n->addint(*one);
+        k = n;
+        if ((n->as_int() & 1) == 0) {
+            coeff = one;
+        } else {
+            coeff = minus_one;
+        }
+    }
+    int j = 1;
+    for (int i = 3; i < 2*k->as_int(); i = i + 2)
+    {
+        j = j * i;
+    }
+    coeff = mulnum(coeff, integer(j));
+    if (arg_->is_positive()) {
+        return div(mul(coeff, sqrt(pi)), pow(i2, n));
+    } else {
+        return div(mul(pow(i2, n), sqrt(pi)), coeff);
+    }
+}
+RCP<const Basic> beta(const RCP<const Basic> &x, const RCP<const Basic> &y)
+{
+    // Only special values are being evaluated    
+    if(eq(*add(x,y),*one)) {
+        throw std::runtime_error("Complex Infinity not yet implemented");
+    }
+
+    if (is_a<Integer>(*x)) {
+        RCP<const Integer> x_int = rcp_static_cast<const Integer>(x);
+        if (x_int->is_positive()) {
+            if (is_a<Integer>(*y)) {
+                RCP<const Integer> y_int = rcp_static_cast<const Integer>(y);
+                if (y_int->is_positive()) {
+                    return div(mul(gamma_positive_int(x),gamma_positive_int(y)),gamma_positive_int(add(x,y)));
```

Comment:
`gamma_positive_int` has the same functionality as in `gamma` right? You should either call `gamma_positive_int` from both `gamma` and `beta` or call `gamma` here